src/lib/pubkey/pk_options.h

```diff
@@ -0,0 +1,459 @@
+      Option<"hash", std::string> hash_fn;
+      Option<"prehash", std::optional<std::string>> prehash;
+      Option<"padding", std::string> padding;
+      Option<"context", std::vector<uint8_t>> context;
+      Option<"provider", std::string> provider;
+      Option<"salt size", size_t> salt_size;
+      Option<"use DER", bool> use_der;
+      Option<"deterministic", bool> deterministic_sig;
+      Option<"explicit trailer field", bool> explicit_trailer_field;
+
+      // NOLINTEND(misc-non-private-member-variables-in-classes)
+
+      auto all_options() const {
+         return std::tie(public_key,
+                         private_key,
+                         rng,
+                         hash_fn,
+                         prehash,
+                         padding,
+                         context,
+                         provider,
+                         salt_size,
+                         use_der,
+                         deterministic_sig,
+                         explicit_trailer_field);
+      }
+};
+
+}  // namespace detail
+
+class BOTAN_UNSTABLE_API PK_Signature_Options final : public Options<detail::PK_Signature_Options_Container> {
+   public:
+      using Options::Options;
+
+   public:
+      [[nodiscard]] auto public_key() { return take(options().public_key); }
+
+      [[nodiscard]] auto private_key() { return take(options().private_key); }
+
+      [[nodiscard]] auto rng() { return take(options().rng); }
+
+      [[nodiscard]] auto hash_function() { return take(options().hash_fn); }
+
+      [[nodiscard]] auto prehash() { return take(options().prehash); }
+
+      [[nodiscard]] auto padding() { return take(options().padding); }
+
+      [[nodiscard]] auto context() { return take(options().context); }
+
+      [[nodiscard]] auto provider() { return take(options().provider); }
+
+      [[nodiscard]] auto salt_size() { return take(options().salt_size); }
+
+      [[nodiscard]] bool using_der_encoded_signature() { return take(options().use_der).or_default(false); }
+
+      [[nodiscard]] bool using_deterministic_signature() { return take(options().deterministic_sig).or_default(false); }
+
+      [[nodiscard]] bool using_explicit_trailer_field() {
+         return take(options().explicit_trailer_field).or_default(false);
+      }
+
+   public:
+      /// It may be acceptable to provide a hash function, for hash-based
+      /// signatures (like SLH-DSA or LMS), but it is not required.
+      /// @throws Invalid_Argument if the provided hash is not acceptable
+      void validate_for_hash_based_signature(std::optional<std::string_view> acceptable_hash = std::nullopt);
+
+      /// This is a convenience helper for algorithms that do not support
+      /// specifying a provider.
+      /// @throws Provider_Not_Found if a provider is set
+      void exclude_provider() {
+         if(auto p = provider().optional()) {
+            throw Provider_Not_Found(product_name(), p.value());
+         };
+      }
```

Comment:
Essentially, this is a convenience construction to replace the ubiquitous sanity check in every algorithm that used to read like:

```C++
if(provider.empty() || provider == "base") {
   // ... do the thing
}
throw Provider_Not_Found();
```